inspect-ast/src/main.rs

```diff
@@ -0,0 +1,232 @@
+use failure::{Error, Fail, ResultExt};
+use regex::Regex;
+use std::{
+    fs::File,
+    io::{self, Read},
+    path::PathBuf,
+    process::exit,
+};
+use structopt::StructOpt;
+use termcolor::{ColorChoice, StandardStream};
+
+use compiler_lib::{
+    asciifile::{AsciiFile, Span, Spanned},
+    ast,
+    context::{self, Context},
+    lexer::{Lexer, TokenKind},
+    parser::Parser,
+    strtab::StringTable,
+    visitor::NodeKind,
+};
+
+#[derive(Debug, Fail)]
+pub enum CliError {
+    #[fail(display = "invalid regular expression")]
+    InvalidRegex,
+    #[fail(display = "cannot open input file {:?}", path)]
+    OpenInput { path: PathBuf },
+    #[fail(display = "cannot decode input file {:?}", path)]
+    Ascii { path: PathBuf },
+}
+
+#[derive(StructOpt)]
+#[structopt(name = "ast-inspector")]
+pub struct CliCommand {
+    /// MiniJava file that should be parsed into an AST.
+    #[structopt(name = "FILE", parse(from_os_str))]
+    input: PathBuf,
+    /// a regular expression used to filter the source code.
+    /// Only matching AST elements will be dumped. Defaults to
+    /// filtering nothing.
+    #[structopt(short = "c", long = "content", default_value = "")]
+    regex_content: Regex,
+    /// a regular expression used to filter the source code.
+    /// Only matching AST elements will be dumped. Defaults to
+    /// filtering nothing.
+    #[structopt(short = "k", long = "node-kind", default_value = "")]
+    regex_node_kind: Regex,
+    /// stop after first regex match
+    #[structopt(short = "1", long = "only-first")]
+    only_first_match: bool,
+}
+
+fn main() {
+    let cmd = CliCommand::from_args();
+
+    if let Err(msg) = do_main(cmd) {
+        exit_with_error(&msg);
+    }
+}
+
+fn do_main(cmd: CliCommand) -> Result<(), Error> {
```

Comment:
refactor this as soon as the *compiler controller* work is done https://comprakt.slack.com/archives/CDEADFB6F/p1542194320005700

Reply:
Yeah. The controller/driver is harder to do than expected :)